### PR TITLE
add reset and release to scope_guard

### PR DESCRIPTION
**Pull request by ericniebler**

### CODE REVIEW

File: include/unifex/scope_guard.hpp [hunk 1]
```diff
@@ -16,27 +16,41 @@
...
-      std::is_nothrow_move_constructible_v<Func>)
+public:
+  scope_guard(Func&& func) noexcept
       : func_((Func &&) func) {}
 
   ~scope_guard() {
+    reset();
+  }
+
+  void dismiss() {
```
Comment (lewissbaker) [inline include/unifex/scope_guard.hpp:40]: noexcept?

File: include/unifex/scope_guard.hpp [hunk 2]
```diff
@@ -16,27 +16,41 @@
...
+    dismissed_ = true;
+  }
+
+  void reset() {
```
Comment (lewissbaker) [inline include/unifex/scope_guard.hpp:44]: `noexcept`?

File: include/unifex/scope_guard.hpp [hunk 3]
```diff
@@ -16,27 +16,41 @@
...
     static_assert(noexcept(((Func &&) func_)()));
-    ((Func &&) func_)();
+    if (!std::exchange(dismissed_, true)) {
+      ((Func &&) func_)();
+    }
   }
```
Comment (kirkshoop) [inline include/unifex/scope_guard.hpp:49]: should dismiss/reset/~scope_guard be marked noexcept?

File: include/unifex/scope_guard.hpp (hunk 2 above)
Comment (lewissbaker) [reply include/unifex/scope_guard.hpp:44]: The name `reset()` doesn't really scream "run this now" to me when I read it at the call-site. I suspect there would be occasions where I would mistake `.reset()` for doing the same thing as `.dismiss()`.

Comment (ericniebler) [reply include/unifex/scope_guard.hpp:44]: Maybe `dismiss` should be `release` to more strongly evoke the `reset`/`release` of standard smart pointers.

File: include/unifex/scope_guard.hpp (hunk 3 above)
Comment (ericniebler) [reply include/unifex/scope_guard.hpp:49]: Well, not the destructor. It's implicitly `noexcept`.

**Pull request merged**